ecs_taskdefinition module fixes and improvements

kaczynskid opened the pull request:
##### Issue Type:

 - Feature Pull Request

##### Plugin Name:

cloud/amazon/ecs_taskdefinition

##### Ansible Version:

```
ansible 2.1.0 (devel 2db3f290ba) last updated 2016/02/24 10:06:09 (GMT +200)
  lib/ansible/modules/core: (detached HEAD 7162623e86) last updated 2016/02/24 10:09:54 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD ecs_servic) last updated 2016/02/24 12:04:25 (GMT +200)
  config file = 
  configured module search path = Default w/o overrides
```

##### Summary:

Module will now compare whole task definition model and update it if any difference found, as opposed to just matching the task name and status - fixes #1169

Module will now fail with upstream Amazon error message when describing the task will fail, as opposed to swallowing the exception and returning None

robynbergeron commented:
Thanks @kaczynskid. @Java1Guy please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

Java1Guy commented:
shipit

robynbergeron commented:
Thanks again to @kaczynskid for this PR, and thanks @Java1Guy for reviewing. Marking for inclusion.

mattjbray commented on cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -120,14 +122,35 @@ def __init__(self, module):
-        except botocore.exceptions.ClientError:
-            return None
+        except botocore.exceptions.ClientError, e:
+            self.module.fail_json(msg="Can't describe task - "+str(e))
```
This exception is raised when trying to describe a task that doesn't exist. Failing here means we can't create new task definitions with this module.

I think something like this should work:

``` python
def describe_task(self, task_name):
    try:
        families_response = self.ecs.list_task_definition_families()
        if task_name not in families_response['families']:
            return None

        response = self.ecs.describe_task_definition(taskDefinition=task_name)
        return response['taskDefinition']
    except botocore.exceptions.ClientError, e:
        self.module.fail_json(msg="Can't describe task - "+str(e))
```

mattjbray commented on cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -120,14 +122,35 @@ def __init__(self, module):
+
+            for mount_point in container['mountPoints']:
+                if 'readOnly' not in mount_point:
+                    mount_point['readOnly'] = 'false'
```
`readOnly` must be a boolean[1]. This causes the following error:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: Invalid type for parameter containerDefinitions[0].mountPoints[0].readOnly, value: false, type: <type 'str'>, valid types: <type 'bool'>
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "parsed": false}
```

The fix is:

``` python
mount_point['readOnly'] = False
```

[1] http://boto3.readthedocs.org/en/latest/reference/services/ecs.html#ECS.Client.register_task_definition

kaczynskid commented:
@mattjbray thx for the review and corrections; I also fixed lists comparison by sotring them; can you have a look if it works for you now?

ryanwalls commented on cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -120,14 +122,58 @@ def __init__(self, module):
+            container['mountPoints'].sort(key=lambda k: k['sourceVolume'])
+            container['portMappings'].sort(key=lambda k: k['hostPort'])
+            container['volumesFrom'].sort(key=lambda k: k['sourceContainer'])
+            container['links'].sort()
```
Blows up if links not defined.  Need to check.

```
            if 'links' in container:
                container['links'].sort()
```

ryanwalls commented on cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -120,14 +122,58 @@ def __init__(self, module):
+            container['volumesFrom'].sort(key=lambda k: k['sourceContainer'])
+            container['links'].sort()
+
+        existingVolumes = sorted(expected['volumes'], key=lambda k: k['name'])
```
This should be `existing['volumes']`

ryanwalls replied:
And you need to check that `volumes` exists.

e.g. 

```


        if existing['volumes'] != None:
            existingVolumes = sorted(existing['volumes'], key=lambda k: k['name'])
        if expected['volumes'] != None:
            expectedVolumes = sorted(expected['volumes'], key=lambda k: k['name'])
```

ryanwalls commented:
And it is also creating a new revision for me every time.

kaczynskid commented:
Thanks @daniel-rhoades @ryanwalls for the review. I added more null checks and refactored comparison function to hopefuly better prepare both amazon data and module data for comparison. Hope it works better now.

I don't get the new revision every time - that was the piont of this change. If you still do, it might meant that I still need to refine data preparation for comparison. It might help if you could write the task definitions you use - I might have missed some properties or sth.

daniel-rhoades commented:
Here is the complete project I am using to test it: https://github.com/daniel-rhoades/hippo-provisioning-template

Its quite involved, but reasonably well documented and structured.  If you have issues with it then I'll create a simple example project

daniel-rhoades commented on cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -136,42 +136,41 @@ def describe_task(self, task_name):
+                # sort lists properties for comaprison
+                container['environment'].sort(key=lambda k: k['name'])
+                container['mountPoints'].sort(key=lambda k: k['sourceVolume'])
+                container['portMappings'].sort(key=lambda k: k['hostPort'])
```
Goes boom here unfortunately:

```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/home/vagrant/.ansible/tmp/ansible-tmp-1458082875.83-119339828575258/ecs_taskdefinition", line 2592, in <module>
    main()
  File "/home/vagrant/.ansible/tmp/ansible-tmp-1458082875.83-119339828575258/ecs_taskdefinition", line 234, in main
    task_mgr.is_matching_task(module.params, existing)):
  File "/home/vagrant/.ansible/tmp/ansible-tmp-1458082875.83-119339828575258/ecs_taskdefinition", line 163, in is_matching_task
    container['portMappings'].sort(key=lambda k: k['hostPort'])
  File "/home/vagrant/.ansible/tmp/ansible-tmp-1458082875.83-119339828575258/ecs_taskdefinition", line 163, in <lambda>
    container['portMappings'].sort(key=lambda k: k['hostPort'])
KeyError: 'hostPort'

fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "ecs_taskdefinition"}, "parsed": false}
```

kaczynskid replied:
Ah! `containerPort` is the mandatory one, not `hostPort` :)

kaczynskid commented:
Thanks @daniel-rhoades for sharing your code. It helped me to understand different use cases. I went to boto3 docs and through a whole `task_definition` model in `is_matching_task()` now. Hopefuly we're getting there :)

resmo commented:
Submitter @kaczynskid, would you mind squash the commits into one?

Not sure about the state of this PR. Asking maintainer @Java1Guy. Is it still ready to ship?

daniel-rhoades commented:
Will test it this evening through my use case

daniel-rhoades commented:
@kaczynskid still no joy.

```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/home/vagrant/.ansible/tmp/ansible-tmp-1458560515.76-50348400278839/ecs_taskdefinition", line 2630, in <module>
    main()
  File "/home/vagrant/.ansible/tmp/ansible-tmp-1458560515.76-50348400278839/ecs_taskdefinition", line 272, in main
    task_mgr.is_matching_task(module.params, existing)):
  File "/home/vagrant/.ansible/tmp/ansible-tmp-1458560515.76-50348400278839/ecs_taskdefinition", line 207, in is_matching_task
    expectedVolumes = list(expected['volumes']) if 'volumes' in expected else []
TypeError: 'NoneType' object is not iterable

fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "ecs_taskdefinition"}, "parsed": false}
```

I don't use volumes for my task definitions, hence the likely source of the error.

kaczynskid commented:
@daniel-rhoades: That's what the `if` sgould handle there. But it seems `list(expected['volumes'])` is evaluated anyway :/ Can you try `list(expected['volumes'] if 'volumes' in expected else [])` instead (same for `existing`)?

@resmo I'll squish it when I'll finish debugging and fixing stuff with @daniel-rhoades

daniel-rhoades commented:
@kaczynskid: Same error.  Although the error goes away with these lines in replacement:

```
existingVolumes = list(existing['volumes']) if existing['volumes'] else []
expectedVolumes = list(expected['volumes']) if expected['volumes'] else []
```

`volumes` was probably already there as a Map key, but with an empty value, this might have been the cause of the previous behaviour.

However, even with those replacement lines we are back to square one, in that it will create a new Task Definition each time.  Still its better behaviour than what is in devel branch at the moment.

I'll have a look through the code, but Python isn't my specialty.

gregdek commented:
Thanks @kaczynskid. To the current maintainers, @Java1Guy please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

daniel-rhoades commented on cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -120,14 +122,95 @@ def __init__(self, module):
+                for list_prop in simple_list_props:
+                    if list_prop not in container:
+                        container[list_prop] = []
+                    else:
```
I've been hitting a problem with my entryPoint commands.  Basically, in my playbook I have:

```
entryPoint:
            - "bash"
            - "-c"
```

It needs to be in that order, but I think because of the sort here it is getting sorted to:

```
entryPoint:
            - "-c"
            - "bash"
```

Which causes the container not to start (Docker can't find the command "-c")

Suggest changing line 170 to:

```
elif list_prop != 'entryPoint':
```

kaczynskid replied:
You are right. It makes no sense to sort `command` and `entryPoint` at all.

jianyuan commented:
In a [recent AWS update](http://aws.amazon.com/about-aws/whats-new/2016/04/aws-cloudformation-updates-amazon-ec2-amazon-ecs-aws-lambda-and-amazon-gamelift-resource-support/), there's new support for additional ECS container definition properties.

As such, the `is_matching_task` method has to be updated.

[Reference here](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-containerdefinitions.html).

jianyuan commented on cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -120,14 +122,95 @@ def __init__(self, module):
 
     def describe_task(self, task_name):
         try:
+            families_response = self.ecs.list_task_definition_families()
```
This endpoint is paginated.

Fix is to use the paginator like so:

``` python
            paginator = self.ecs.get_paginator('list_task_definition_families')
            for families_response in paginator.paginate():
                if task_name in families_response['families']:
                    response = self.ecs.describe_task_definition(
                        taskDefinition=task_name
                    )
                    return response['taskDefinition']
            else:
                return None
```

pvwichen commented on cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -120,14 +122,95 @@ def __init__(self, module):
+
+        # prepare volumes for comaprison
+        existingVolumes = list(existing['volumes']) if 'volumes' in existing else []
+        expectedVolumes = list(expected['volumes']) if 'volumes' in expected else []
```
The 'volumes' key may be present in the `expected` dictionary with the value of `NoneType`.

Should line 206 be:
`expectedVolumes = list(expected['volumes']) if 'volumes' in expected and expected['volumes'] is not None else []`

pvwichen commented on cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -120,14 +122,95 @@ def __init__(self, module):
+                        'readOnly': False
+                    },
+                    'portMappings': {
+                        'protocol': 'tcp'
```
When using dynamic ports, the existing values are returned as:

```
"portMappings": [{
                "containerPort": 80,
                "hostPort": 0,
                "protocol": "tcp"
            }],
```

Should the `complex_list_defaults` entry for `portMappings` be:

```
                  'portMappings': {
                        'protocol': 'tcp',
                        'hostPort': 0
                    },
```

kaczynskid commented:
Thank you all guys for great commens and fix suggestions - I applied them all.

I also rebased this branch against stable2.1 branch, bu I cannot edit the pull request. Can this request be edited by someone to merge into stable2.1? Or should I close this one and create a separate pull request? Or chould I rebase this against current devel?

kaczynskid commented:
OK, so I found out that pull request target branch cannot be changed. Therefore I rebased the changes against current develop and squished the commits. This is still compatible with `stable-2.1` and I'd love to see it merged there as well and I'll create a separate pull request for that when this is merged.

igreg commented:
@kaczynskid I have noticed that it is not possible to remove an ECS task definition by supplying family and revision number. When it is the case, `task_to_describe` is set to `familyname:revision` and the plugin uses this to find a matching task which fails because `describe_task()` expect `task_name` to be a family.

Actually, I think even submitting the ARN of the task to remove will fail as well.

The reason it fails to remove the task because this line 292:

``` python
    elif module.params['state'] == 'absent':
        if not existing:
            pass
```

Do you think you could fix it?

Edit: this is fixed in the other PR (#2357) improving the `ecs_taskdefinition` module. A reason to favour it in place of this PR?

kaczynskid commented:
@igreg Thx for pointing PR #2357. I could fix the removal, but I like the other PR implementation better. I'm not a python programmer, so I just fixed whatever I needed the best I could, while @rmorlok seems to know better what he's doing ;) So I'd myslef favor his PR over mine here.

@vmalaga You probbably also should go and see PR #2357 instead :)

kaczynskid commented:
Closing in favor of PR #2357
